src/lib/block-batch.js

```diff
@@ -1,35 +1,42 @@
 const MAX_BYTES_BETWEEN = 1024 * 1024 * 2
 const MAX_BATCH_SIZE = 10
 
-export class BlockBatch {
-  /** @type {number[]} */
-  #offsets = []
+/**
+ * @typedef {import('multiformats').CID} CID
+ * @typedef {{ carCid: CID, blockCid: CID, offset: number }} BatchItem
+ * @typedef {{ add: (i: BatchItem) => void, next: () => BatchItem[] }} BlockBatcher
+ */
 
-  /**
-   * Add an offset to the batch
-   * @param {number} offset
-   */
-  add (offset) {
-    this.#offsets.push(offset)
+/**
+ * Batcher for blocks in CARs. Batches are grouped by CAR CID and blocks are
+ * returned in batches in the order they were inserted.
+ * @implements {BlockBatcher}
+ */
+export class OrderedCarBlockBatcher {
+  /** @type {BatchItem[]} */
+  #queue = []
+
+  /** @param {BatchItem} item */
+  add (item) {
+    this.#queue.push(item)
   }
 
   next () {
-    const offsets = this.#offsets // .sort((a, b) => a - b)
-    if (!offsets.length) return
+    const queue = this.#queue
+    if (!queue.length) return []
     const batch = []
-    let last = offsets[0]
+    let last = queue[0]
     while (true) {
-      const offset = offsets.shift()
-      if (!offset) break
-      if (offset - last >= MAX_BYTES_BETWEEN) {
-        offsets.unshift(offset) // not in this batch, return to pile
+      const item = queue.shift()
+      if (!item) break
+      if (item.carCid.toString() !== last.carCid.toString() || item.offset - last.offset >= MAX_BYTES_BETWEEN) {
+        queue.unshift(item) // not in this batch, return to pile
```

Comment:
minor but could skip a gozzy shift/unshift operation and invert the logic to make it more positive... like test for the condition we want to be true

```suggestion
```suggestion
      const item = queue.at(0)
      if (!item) break
      if (item.carCid.toString() === last.carCid.toString() && item.offset - last.offset < MAX_BYTES_BETWEEN) {
        batch.push(queue.shift(item)) // winner! remove from queue and add to item
``````